examples/deutsch.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2018 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Demonstrates Deutsch's algorithm.
+
+Deutsch's algorithm is one of the simplest demonstrations of quantum parallelism
+and interference. It takes a black-box oracle implementing a Boolean function
+f(x), and determines whether f(0) and f(1) have the same parity using just one
+query.  This version of Deutsch's algorithm is a simplified and improved version
+from Nielsen and Chuang's textbook.
+
+=== REFERENCE ===
+
+Deutsch, David. "Quantum theory, the Church-Turing Principle and the universal
+quantum computer." Proc. R. Soc. Lond. A, 400:97, 1985.
+
+Nielsen, Michael A. and Isaac L. Chuang. "Quantum Computation and Quantum
```

Comment:
A user might not have the book handy. We could reference a link. The Wikipedia page describing this algorithm: https://en.wikipedia.org/wiki/Deutsch%E2%80%93Jozsa_algorithm is correct. We could link to it. If you are not comfortable with that, there are lecture notes online we could link to.

Reply:
Definitely link to wikipedia. Much more practical.